src/v/config/property.h

```diff
@@ -953,5 +963,118 @@ class hidden_when_default_property : public property<T> {
         return this->value() == this->default_value();
     }
 };
+namespace detail {
+
+template<typename P>
+concept Property = requires() {
+    std::derived_from<P, base_property>;
+    typename P::underlying;
+};
+
+template<typename T>
+concept Array = !(std::is_same_v<std::decay_t<T>, ss::sstring>
+                  || std::is_same_v<std::decay_t<T>, std::string>)
+                && is_collection<T>;
+
+template<typename T>
+concept OptArray = reflection::is_std_optional<T>
+                   && Array<typename T::value_type>;
+
+} // namespace detail
+
+template<detail::Property P>
+class enterprise : public P {
+    template<typename T, typename = void>
+    struct value_type {
+        using type = T;
+    };
+    template<detail::Array T>
+    struct value_type<T, std::void_t<typename T::value_type>> {
+        using type = T::value_type;
+    };
+    template<reflection::is_std_optional T>
+    struct value_type<T, std::void_t<typename T::value_type>> {
+        using type = T::value_type;
+    };
+
+    using T = typename P::underlying;
+    using ValT = typename value_type<T>::type;
+    using val_container_t = std::vector<ValT>;
+    using restrict_check_t = std::function<bool(const ValT&)>;
+
+public:
+    template<typename R, typename... Args>
+    requires std::ranges::range<R>
+               && std::is_same_v<std::ranges::range_value_t<R>, ValT>
+    enterprise(config_store& conf, R restrict, Args&&... args)
```

Comment:
nit: `restrict` is a restricted word [sic].
While for c++ the above is not true, github still thinks it is worth being hightlighted. How about `restricted`, so it doesn't stand out?

Reply:
yeah you're the second person to mention that (c99 keyword with absolutely wacky semantics if you're interested). I'll change it.

Reply:
done